contracts.py

```diff
@@ -1,15 +1,22 @@
 import json
+import os
 import pathlib
 import re
 
 from first import first
 from mkdocs.plugins import BasePlugin
 
 
-class_path_pat = re.compile(r"\n[\s]*!!!\s(.*)\n")
+backends_pat = re.compile(r"\n[\s]*!!!\sbackend:(.*)\n")
+contracts_pat = re.compile(r"\n[\s]*!!!\scontract:(.*)\n")
```

Comment:
I feel like this is a stupid (or at least ignorant) question, but how does the docstring markdown get generated with `!!! contract:`

Reply:
That's a good question!  Which should probably be in the readme… 

I'll summarise here and add it to the readme in another PR.

There's couple of pieces to this…

`python -m cohortextractor generate_docs` creates a JSON file with details (name, dotted path, docstring, etc) for each of the table contracts.  It also includes the tables grouped by backend.

Mkdocs plugins follow an evented (event-driven?) pattern where plugins implement one of the `on_page_*` methods to make changes to the relevant event (so markdown for us).  That method recieves the content from each markdown file in the docs repo.  When it find somethings which matches one of the patterns above (eg a line which looks like `!!! contract:cohortextractor2.concepts.PatientDemographics`) it uses the captured pattern (`cohortextractor2.concepts.PatientDemographics`) to look up the appropriate section of the JSON file generated by cohortextractor-v2 and pull the relevant bits of info (just the name and docstring currently) for rendering into markdown.

Reply:
OK, right, so I got all of that from my reading of the PRs and mkdocs plugin docs.  I get what's going on, but how does the line like `!!! contract:cohortextractor2.concepts.PatientDemographics` get generated?  How does that class (or how does mkdocs) know it's a contract and generate the markdown line with `!!! contract:...`?

Reply:
Wait - docs repo.  That's it, isn't it :)

Reply:
Got it! https://github.com/opensafely/documentation/pull/505/